modules/jooby-metrics/src/main/java/io/jooby/metrics/MetricsModule.java

```diff
@@ -0,0 +1,277 @@
+   * @param metric A metric object.
+   * @param <M> Metric type.
+   * @return This metrics module.
+   */
+  public <M extends Metric> MetricsModule metric(final String name, final Class<M> metric) {
+    metricClasses.put(name, metric);
+    return this;
+  }
+
+  /**
+   * Append a health check to the {@link HealthCheckRegistry}. This call is identical to
+   * {@link HealthCheckRegistry#register(String, HealthCheck)}.
+   *
+   * @param name Name of the check.
+   * @param check A check object.
+   * @return This metrics module.
+   */
+  public MetricsModule healthCheck(final String name, final HealthCheck check) {
+    healthChecks.put(name, check);
+    return this;
+  }
+
+  /**
+   * Append a health check to the {@link HealthCheckRegistry}. The metric will be resolved by Guice.
+   * This call is identical to {@link HealthCheckRegistry#register(String, HealthCheck)}.
+   *
+   * @param name Name of the check.
+   * @param check A check object.
+   * @param <H> {@link HealthCheck} type.
+   * @return This metrics module.
+   */
+  public <H extends HealthCheck> MetricsModule healthCheck(final String name, final Class<H> check) {
+    healthCheckClasses.put(name, check);
+    return this;
+  }
+
+  /**
+   * Append a {@link Reporter} to the {@link MetricRegistry}.
+   *
+   * @param callback Reporter callback.
+   * @return This metrics module.
+   */
+  public MetricsModule reporter(final BiFunction<MetricRegistry, Config, Reporter> callback) {
+    this.reporters.add(requireNonNull(callback, "Callback is required."));
+    return this;
+  }
+
+  /**
+   * Append a {@link Reporter} to the {@link MetricRegistry}.
+   *
+   * @param callback Reporter callback.
+   * @return This metrics module.
+   */
+  public MetricsModule reporter(final Function<MetricRegistry, Reporter> callback) {
+    return reporter((registry, conf) -> callback.apply(registry));
+  }
+
+  @Override
+  public void install(@Nonnull Jooby application) {
+    MetricHandler metricHandler = new MetricHandler();
+    application.route("GET", this.pattern + "/metrics", metricHandler);
+    application.route("GET", this.pattern + "/metrics/:type", metricHandler);
+    application.route("GET", this.pattern + "/healthcheck", new HealthCheckHandler());
+
+    routes.forEach(r -> r.accept(application));
+
+    ServiceRegistry registry = application.getServices();
+
+    registry.putIfAbsent(MetricRegistry.class, metricRegistry);
+    registry.putIfAbsent(HealthCheckRegistry.class, healthCheckRegistry);
+
+    final Set<Reporter> reporters = new HashSet<>();
+
+    application.onStarted(() -> {
+      metrics.forEach((name, instance) -> {
+        registry.putIfAbsent(ServiceKey.key(Metric.class, name), instance);
+        metricRegistry.register(name, instance);
+      });
+
+      metricClasses.forEach((name, clazz) -> {
+        Metric instance = application.require(clazz);
+        registry.putIfAbsent(ServiceKey.key(Metric.class, name), instance);
```

Comment:
Classes must run on application started (after DI is ready) but there is no sense to add them in registry. Lifecycle of them are handle by DI framework.

Reply:
I see the point. Then maybe we should neither add the user code provided instances as well? Their life cycle should be handled by the user code.

Reply:
Yea, but for DI usage what you did here is more clear and we make sure we request the object at time DI is ready. It is like more "safe". We need just these lines:

```
Metric instance = application.require(clazz);
metricRegistry.register(name, instance);
```